pywt/_extensions/c/wt.template.c

```diff
@@ -447,20 +448,131 @@ int CAT(TYPE, _swt_)(TYPE input[], pywt_index_t input_len,
+    // These are boolean values, but MSVC does not have <stdbool.h>
+    int make_temp_input, make_temp_output;
+
+    if (input_info.ndim != output_info.ndim)
+        return 1;
+    if (axis >= input_info.ndim)
+        return 1;
+
+    for (i = 0; i < input_info.ndim; ++i){
+        if (i == axis){
+            if (swt_buffer_length(input_info.shape[i])
+                != output_info.shape[i])
+                return 1;
+        } else {
+            if (input_info.shape[i] != output_info.shape[i])
+                return 1;
+        }
+    }
+
+    make_temp_input = input_info.strides[axis] != sizeof(TYPE);
+    make_temp_output = output_info.strides[axis] != sizeof(TYPE);
+    if (make_temp_input)
+        if ((temp_input = malloc(input_info.shape[axis] * sizeof(TYPE))) == NULL)
+            goto cleanup;
+    if (make_temp_output)
+        if ((temp_output = malloc(output_info.shape[axis] * sizeof(TYPE))) == NULL)
+            goto cleanup;
+
+    for (i = 0; i < output_info.ndim; ++i){
+        if (i != axis)
+            num_loops *= output_info.shape[i];
+    }
+
+    for (i = 0; i < num_loops; ++i){
+        size_t j;
+        size_t input_offset = 0, output_offset = 0;
+        const TYPE * input_row;
+        TYPE * output_row;
+
+        // Calculate offset into linear buffer
+        {
+            size_t reduced_idx = i;
+            for (j = 0; j < output_info.ndim; ++j){
+                size_t j_rev = output_info.ndim - 1 - j;
+                if (j_rev != axis){
+                    size_t axis_idx = reduced_idx % output_info.shape[j_rev];
+                    reduced_idx /= output_info.shape[j_rev];
+
+                    input_offset += (axis_idx * input_info.strides[j_rev]);
+                    output_offset += (axis_idx * output_info.strides[j_rev]);
+                }
+            }
+        }
+
+        // Copy to temporary input if necessary
+        if (make_temp_input)
+            for (j = 0; j < input_info.shape[axis]; ++j)
+                // Offsets are byte offsets, to need to cast to char and back
+                temp_input[j] = *(TYPE *)(((char *) input) + input_offset
+                                          + j * input_info.strides[axis]);
+
+        // Select temporary or direct output and input
+        input_row = make_temp_input ? temp_input
+            : (const TYPE *)((const char *) input + input_offset);
+        output_row = make_temp_output ? temp_output
+            : (TYPE *)((char *) output + output_offset);
+
+        // Apply along axis
+        switch (coef){
+        case COEF_APPROX:
+            CAT(TYPE, _swt_a)(input_row, input_info.shape[axis],
+                              wavelet, output_row, output_info.shape[axis],
+                              level);
+            break;
+        case COEF_DETAIL:
+            CAT(TYPE, _swt_d)(input_row, input_info.shape[axis],
+                              wavelet, output_row, output_info.shape[axis],
+                              level);
```

Comment:
A warning here:

    In file included from pywt/_extensions/c/wt.c:12:0:
    pywt/_extensions/c/wt.template.c:555:31: warning: passing argument 1 of ‘double_swt_d’ discards ‘const’ qualifier from pointer target type [enabled by default]
                               level);